File: java/com/google/copybara/transform/metadata/ReferenceMigrator.java
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.transform.metadata;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.copybara.ChangeVisitable.VisitResult.CONTINUE;
import static com.google.copybara.ChangeVisitable.VisitResult.TERMINATE;
import static com.google.copybara.config.SkylarkUtil.check;
import static com.google.copybara.exception.ValidationException.checkCondition;

import com.google.common.base.MoreObjects;
import com.google.common.base.Objects;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.copybara.ChangeVisitable;
import com.google.copybara.TransformWork;
import com.google.copybara.Transformation;
import com.google.copybara.TransformationStatus;
import com.google.copybara.exception.RepoException;
import com.google.copybara.exception.ValidationException;
import com.google.copybara.transform.ExplicitReversal;
import com.google.copybara.transform.IntentionalNoop;
import com.google.copybara.transform.RegexTemplateTokens;
import com.google.copybara.transform.RegexTemplateTokens.Replacer;
import com.google.re2j.Pattern;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import javax.annotation.Nullable;
import net.starlark.java.eval.EvalException;
import net.starlark.java.syntax.Location;

/**
 * Adjusts textual references in  change messages to match the destination.
 *
 */
public class ReferenceMigrator implements Transformation {


  static final int MAX_CHANGES_TO_VISIT = 5000;

  private final RegexTemplateTokens before;
  private final RegexTemplateTokens after;
  private final ImmutableList<String> additionalLabels;

  @Nullable private final Pattern reversePattern;
  private final Location location;

  private final Map<String, String> knownChanges = new HashMap<>();

  ReferenceMigrator(
      RegexTemplateTokens before,
      RegexTemplateTokens after,
      @Nullable Pattern reversePattern,
      ImmutableList<String> additionalLabels,
      Location location) {
    this.before = checkNotNull(before, "before");
    this.after = checkNotNull(after, "after");
    this.additionalLabels = checkNotNull(additionalLabels, "additionalLabels");
    this.reversePattern = reversePattern;
    this.location = checkNotNull(location);
  }

  public static ReferenceMigrator create(
      String before, String after, Pattern forward, @Nullable Pattern backward,
      ImmutableList<String> additionalLabels, Location location) throws EvalException {
    Map<String, Pattern> patterns = ImmutableMap.of("reference", forward);
    RegexTemplateTokens beforeTokens =
        new RegexTemplateTokens(before, patterns, /* repeatedGroups= */ false, location);
    beforeTokens.validateUnused();
    RegexTemplateTokens afterTokens =
        new RegexTemplateTokens(after, patterns, /* repeatedGroups= */ false, location);
    afterTokens.validateUnused();
    check(
        after.lastIndexOf("$1") == -1,
        "Destination format '%s' uses the reserved token '$1'.",
        after);
    return new ReferenceMigrator(beforeTokens, afterTokens, backward, additionalLabels,
        location);
  }

  @Override
  public TransformationStatus transform(TransformWork work) throws ValidationException {
    AtomicReference<ValidationException> thrown = new AtomicReference<>();
    Replacer replacer = before.callbackReplacer(after, (groupValues, template) -> {
        if (groupValues.get(0) != null) {
          try {
            String destinationRef = findChange(groupValues.get(1),
                work.getMigrationInfo().getOriginLabel(),
                work.getMigrationInfo().destinationVisitable());
            if (destinationRef != null) {
              // This will not work for the case where the template was "foo\\$1", if this is an
              // issue, a non-naive implementation might be required.
              return Pattern.compile("[$]1").matcher(template).replaceAll(destinationRef);
            } else {
              return groupValues.get(0);
            }
          } catch (ValidationException exception) {
            thrown.compareAndSet(null, exception);
            return groupValues.get(0);
          }
        }
        return template;
      }, false, false, null);
    String replaced = replacer.replace(work.getMessage());
    if (thrown.get() != null) {
      throw thrown.get();
    }
    if (!replaced.equals(work.getMessage())) {
      work.setMessage(replaced);
    }
    return TransformationStatus.success();
  }

  @Override
  public Transformation reverse() {
    return new ExplicitReversal(IntentionalNoop.INSTANCE, this);
  }

  @Override
  public String describe() {
    return "map_references: " + before + " to " + after;
  }

  @Nullable
  private String findChange(String refBeingMigrated,
      String originLabel,
      ChangeVisitable<?> destinationReader) throws  ValidationException {
    AtomicInteger changesVisited = new AtomicInteger(0);
    ImmutableList<String> originLabels =
        ImmutableList.<String>builder().add(originLabel).addAll(additionalLabels).build();
    checkCondition(destinationReader != null,
        "Destination does not support reading change history.");
    if (knownChanges.containsKey(refBeingMigrated)) {
      return knownChanges.get(refBeingMigrated);
    }
    try {
      destinationReader.visitChangesWithAnyLabel(null, originLabels, (input, labels) -> {
        for (String labelValue : labels.values()) {
          knownChanges.putIfAbsent(labelValue, input.getRef());
          if (labelValue.equals(refBeingMigrated)) {
            return TERMINATE;
          }
        }
        return changesVisited.incrementAndGet() > MAX_CHANGES_TO_VISIT ? TERMINATE : CONTINUE;
      });
      String retVal = knownChanges.get(refBeingMigrated);
      if (reversePattern != null && retVal != null && !reversePattern.matches(retVal)) {
        throw new ValidationException(
            String.format("Reference %s does not match regex '%s'", retVal, reversePattern));
      }
      return retVal;
    } catch (RepoException exception) {
      throw new ValidationException("Exception finding reference.", exception);
    }
  }

  @Override
  public String toString() {
    return MoreObjects.toStringHelper(this)
        .add("before", before)
        .add("after", after)
        .toString();
  }

  @Override
  public boolean equals(Object other) {
    return other instanceof ReferenceMigrator
        && Objects.equal(this.before,
        ((ReferenceMigrator) other).before)
        && Objects.equal(this.after,
        ((ReferenceMigrator) other).after);
  }

  @Override
  public Location location() {
    return location;
  }

  @Override
  public int hashCode() {
    return Objects.hashCode(before, after);
  }
}

File: src/Copybara.Core/Transform/Metadata/ReferenceMigrator.cs
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Immutable;
using System.Text.RegularExpressions;
using Copybara.Common;
using Copybara.Exceptions;
using Copybara.Revision;
using Copybara.TemplateToken;
using Copybara.Transform;
using Starlark.Eval;
using Starlark.Syntax;
using StarlarkRt = Starlark.Eval.Starlark;

namespace Copybara.Transform.Metadata;

/// <summary>Adjusts textual references in change messages to match the destination.</summary>
public class ReferenceMigrator : ITransformation
{
    internal const int MaxChangesToVisit = 5000;

    private readonly RegexTemplateTokens _before;
    private readonly RegexTemplateTokens _after;
    private readonly ImmutableArray<string> _additionalLabels;
    private readonly Regex? _reversePattern;
    private readonly Location _location;

    private readonly Dictionary<string, string> _knownChanges = new();

    internal ReferenceMigrator(
        RegexTemplateTokens before,
        RegexTemplateTokens after,
        Regex? reversePattern,
        ImmutableArray<string> additionalLabels,
        Location location)
    {
        _before = Preconditions.CheckNotNull(before);
        _after = Preconditions.CheckNotNull(after);
        _additionalLabels = additionalLabels;
        _reversePattern = reversePattern;
        _location = Preconditions.CheckNotNull(location);
    }

    public static ReferenceMigrator Create(
        string before,
        string after,
        Regex forward,
        Regex? backward,
        ImmutableArray<string> additionalLabels,
        Location location)
    {
        var patterns = new Dictionary<string, Regex> { ["reference"] = forward };
        var beforeTokens = new RegexTemplateTokens(before, patterns, repeatedGroups: false, location);
        beforeTokens.ValidateUnused();
        var afterTokens = new RegexTemplateTokens(after, patterns, repeatedGroups: false, location);
        afterTokens.ValidateUnused();
        if (after.LastIndexOf("$1", StringComparison.Ordinal) != -1)
        {
            throw StarlarkRt.Errorf(
                "Destination format '{0}' uses the reserved token '$1'.", after);
        }
        return new ReferenceMigrator(beforeTokens, afterTokens, backward, additionalLabels, location);
    }

    public TransformationStatus Transform(TransformWork work)
    {
        ValidationException? thrown = null;
        var replacer = _before.CallbackReplacer(
            _after,
            new Callback(this, work, () => thrown, ex => thrown ??= ex),
            firstOnly: false,
            multiline: false,
            patternsToIgnore: null);
        string replaced = replacer.Replace(work.GetMessage());
        if (thrown != null)
        {
            throw thrown;
        }
        if (!replaced.Equals(work.GetMessage()))
        {
            work.SetMessage(replaced);
        }
        return TransformationStatus.Success();
    }

    private sealed class Callback : RegexTemplateTokens.IAlterAfterTemplate
    {
        private readonly ReferenceMigrator _owner;
        private readonly TransformWork _work;
        private readonly Func<ValidationException?> _getThrown;
        private readonly Action<ValidationException> _setThrown;

        public Callback(
            ReferenceMigrator owner,
            TransformWork work,
            Func<ValidationException?> getThrown,
            Action<ValidationException> setThrown)
        {
            _owner = owner;
            _work = work;
            _getThrown = getThrown;
            _setThrown = setThrown;
        }

        public string Alter(IReadOnlyDictionary<int, string> groupValues, string template)
        {
            groupValues.TryGetValue(0, out var whole);
            if (whole != null)
            {
                try
                {
                    groupValues.TryGetValue(1, out var refValue);
                    string? destinationRef = _owner.FindChange(
                        refValue!,
                        _work.GetMigrationInfo().GetOriginLabel(),
                        _work.GetMigrationInfo().DestinationVisitable());
                    if (destinationRef != null)
                    {
                        // This will not work for the case where the template was "foo\\$1", if this
                        // is an issue, a non-naive implementation might be required.
                        return Regex.Replace(template, "[$]1", destinationRef.Replace("$", "$$"));
                    }
                    return whole;
                }
                catch (ValidationException exception)
                {
                    if (_getThrown() == null)
                    {
                        _setThrown(exception);
                    }
                    return whole;
                }
            }
            return template;
        }
    }

    public ITransformation Reverse() =>
        new ExplicitReversal(IntentionalNoop.Instance, this);

    public string Describe() => "map_references: " + _before + " to " + _after;

    private string? FindChange(
        string refBeingMigrated,
        string originLabel,
        IChangeVisitable<IRevision>? destinationReader)
    {
        int changesVisited = 0;
        var originLabels = new List<string> { originLabel };
        originLabels.AddRange(_additionalLabels);
        ValidationException.CheckCondition(
            destinationReader != null, "Destination does not support reading change history.");
        if (_knownChanges.TryGetValue(refBeingMigrated, out var known))
        {
            return known;
        }
        try
        {
            destinationReader!.VisitChangesWithAnyLabel(
                null,
                originLabels,
                new LabelVisitor(this, refBeingMigrated, () => ++changesVisited));
            _knownChanges.TryGetValue(refBeingMigrated, out var retVal);
            if (_reversePattern != null && retVal != null && !IsFullMatch(_reversePattern, retVal))
            {
                throw new ValidationException(
                    $"Reference {retVal} does not match regex '{_reversePattern}'");
            }
            return retVal;
        }
        catch (RepoException exception)
        {
            throw new ValidationException("Exception finding reference.", exception);
        }
    }

    private static bool IsFullMatch(Regex regex, string input)
    {
        var m = regex.Match(input);
        return m.Success && m.Index == 0 && m.Length == input.Length;
    }

    private sealed class LabelVisitor : IChangesLabelVisitor
    {
        private readonly ReferenceMigrator _owner;
        private readonly string _refBeingMigrated;
        private readonly Func<int> _incrementVisited;

        public LabelVisitor(
            ReferenceMigrator owner, string refBeingMigrated, Func<int> incrementVisited)
        {
            _owner = owner;
            _refBeingMigrated = refBeingMigrated;
            _incrementVisited = incrementVisited;
        }

        public VisitResult Visit(
            Change<IRevision> input, IReadOnlyDictionary<string, string> matchedLabels)
        {
            foreach (var labelValue in matchedLabels.Values)
            {
                if (!_owner._knownChanges.ContainsKey(labelValue))
                {
                    _owner._knownChanges[labelValue] = input.Ref;
                }
                if (labelValue.Equals(_refBeingMigrated))
                {
                    return VisitResult.Terminate;
                }
            }
            return _incrementVisited() > MaxChangesToVisit
                ? VisitResult.Terminate
                : VisitResult.Continue;
        }
    }

    public override string ToString() => $"ReferenceMigrator{{before={_before}, after={_after}}}";

    public override bool Equals(object? other) =>
        other is ReferenceMigrator o
        && Equals(_before, o._before)
        && Equals(_after, o._after);

    public Location Location() => _location;

    public override int GetHashCode() => HashCode.Combine(_before, _after);
}
